File: function.r

kaggle_fun=function(y,pred_y){

	n=length(y)
	pred_y2=round(pred_y)
	v1 = log(y+1)-log(pred_y2+1)
	v2 = v1^2
	v3 = sum(v2)
	v4 = sqrt(v3/n)

	return(v4)
}

work.train.test.map.fun2=function(n,amount){
	#n=nrow(train.x)
	num=c(1:n)	
	if( n*0.7>=amount){
		n1 = sample(num,amount)
	}else if( n*0.7<amount){
		n1 = sample(num,n*0.7)		
	}	
	n2=num[-n1]
	return(list(n1,n2))
}

xgb.fun=function(train,test3,nro,eta=0.1,md=10,cb=0.5){

	clf <- xgb.train(params=list(  objective="reg:linear", 
                               booster = "gbtree",
                               eta=eta, 
                               max_depth=md, 
                               colsample_bytree=cb) ,
                	data = xgb.DMatrix(
				data=data.matrix(
				train[,c(8:ncol(train)),with=FALSE]),
                	label=data.matrix(train[,.(log.due)])
				,missing=NA), 
				nrounds =nro,
                 	verbose = 1,
                 	print_every_n=5,
                 early_stopping_rounds    = 10,
                 maximize            = FALSE,
                 eval_metric='rmse'
	)

	pred1<-predict(clf,xgb.DMatrix(data.matrix(
					train[,c(8:ncol(train)),with=FALSE]),
					missing=NA))
	pred1.log = expm1( pred1 )
	value1 = kaggle_fun(train$Demanda_uni_equil,pred1.log)

	pred2<-predict(clf,xgb.DMatrix(data.matrix(
					test3[,c(8:ncol(test3)),with=FALSE]),
					missing=NA))
	pred2.log = expm1( pred2 )
	pred2.log[pred2.log<0]=0
	value2 = kaggle_fun(test3$Demanda_uni_equil,pred2.log)
#--------------------------------------------------------------------------
	return( list( c(value1,value2 ) , clf ) )
}

work.train.data.fun=function(test2){

	temp	 = work.train.test.map.fun2( nrow(test2),2000000 )

	train.num=temp[[1]]
	test.num=temp[[2]]

	train=test2[train.num,]
	test3=test2[test.num,]
	return( list(train,test3) )
}
find.map.fun=function(x){
	na.map = apply(x,1,is.na)
	row.na = apply( t(na.map) , 1 , sum )
	#沒有na的data
	map0 = (  row.na == 0 )
	sum(map0)
	return(map0)
}
na.become.0.fun=function(test,i){
	map = is.na( test[,c(i),with=FALSE] )
	test[map[,1],c(i)]<-0
	return(test)
}


work.var.fun=function(main.train.x){

	#main.train.x$log.vuh	= log1p( main.train.x$Venta_uni_hoy )	#sale
	main.train.x$log.vh  = log1p( main.train.x$Venta_hoy )		#sale money
	#main.train.x$log.dup = log1p( main.train.x$Dev_uni_proxima )	#return
	main.train.x$log.due = log1p( main.train.x$Demanda_uni_equil )#adj
	gc()
	#--------------------------------------------------------------------------------------------
	mean.due.age = main.train.x[, .(mean.due.age = mean(log.due)), by = .(Agencia_ID)]
	mean.due.can = main.train.x[, .(mean.due.can = mean(log.due)), by = .(Canal_ID)]
	mean.due.rut = main.train.x[, .(mean.due.rut = mean(log.due)), by = .(Ruta_SAK)]
	mean.due.cli = main.train.x[, .(mean.due.cli = mean(log.due)), by = .(Cliente_ID)]
	mean.due.pro = main.train.x[, .(mean.due.pro = mean(log.due)), by = .(Producto_ID)]
	#--------------------------------------------------------------------------------------------
	mean.vh.age = main.train.x[, .(mean.vh.age = mean(log.vh)), by = .(Agencia_ID)]
	#--------------------------------------------------------------------------------------------
	#--------------------------------------------------------------------------------------------
	mean.due.pa	= main.train.x[, .(mean.due.pa= mean(log.due)), by = .(Producto_ID, Agencia_ID)]
	mean.due.pr	= main.train.x[, .(mean.due.pr= mean(log.due)), by = .(Producto_ID, Ruta_SAK)]
	mean.due.pcli	= main.train.x[, .(mean.due.pcli= mean(log.due)), by = .(Producto_ID, Cliente_ID)]
	mean.due.pcan	= main.train.x[, .(mean.due.pcan= mean(log.due)), by = .(Producto_ID, Canal_ID)]

	mean.due.pca	= main.train.x[, .(mean.due.pca= mean(log.due)), by = .(Producto_ID, Cliente_ID, Agencia_ID)]
	#--------------------------------------------------------------------------------------------

	mean.due.acrcp	= main.train.x[, .(mean.due.acrcp= mean(log.due)), 
		by = .(Agencia_ID , Canal_ID , Ruta_SAK , Cliente_ID , Producto_ID)]

	sd.due.acrcp	= main.train.x[, .(sd.due.acrcp= sd(log.due)), 
		by = .(Agencia_ID , Canal_ID , Ruta_SAK , Cliente_ID , Producto_ID)]

	#--------------------------------------------------------------------------------------------
	return( list(
			mean.due.age,
			mean.due.can,
			mean.due.rut ,
			mean.due.cli ,
			mean.due.pro ,
			mean.due.pa,
			mean.due.pr,
			mean.due.pcli,
			mean.due.pcan,
			mean.due.pca,	
			mean.vh.age,
			mean.due.acrcp,
			sd.due.acrcp
	) )
}

work.model.fun2=function(main.train.y,
                         n,sample.amount,
                         mean.due.age,
                         mean.due.can,
                         mean.due.rut ,
                         mean.due.cli ,
                         mean.due.pro ,
                         mean.due.pa,
                         mean.due.pr,
                         mean.due.pcli,
                         mean.due.pcan,
                         mean.due.pca,
                         mean.vh.age,
                         mean.due.acrcp,
                         sd.due.acrcp){
  
  set.seed(100)
  #n=1000000
  num = sample( c( 1:sample.amount ) ,n )
  
  test = main.train.y[num,.(	Agencia_ID , Canal_ID , Ruta_SAK , 
                             Cliente_ID , Producto_ID , log.due , 
                             Demanda_uni_equil )]	
  #--------------------------------------------------------------------------------------------
  test = merge(test , mean.due.age, all.x = TRUE, by = c("Agencia_ID"))
  test = merge(test , mean.due.can, all.x = TRUE, by = c("Canal_ID"))
  test = merge(test , mean.due.rut, all.x = TRUE, by = c("Ruta_SAK"))
  test = merge(test , mean.due.cli, all.x = TRUE, by = c("Cliente_ID"))
  #test = merge(test , mean.due.pro, all.x = TRUE, by = c("Producto_ID"))
  #--------------------------------------------------------------------------------------------
  #--------------------------------------------------------------------------------------------
  #--------------------------------------------------------------------------------------------
  test = merge(test , mean.due.pa, all.x = TRUE, by = c("Producto_ID", "Agencia_ID"))
  test = merge(test , mean.due.pr, all.x = TRUE, by = c("Producto_ID", "Ruta_SAK"))
  test = merge(test , mean.due.pcli, all.x = TRUE, by = c("Producto_ID", "Cliente_ID"))
  test = merge(test , mean.due.pcan, all.x = TRUE, by = c("Producto_ID", "Canal_ID"))
  #--------------------------------------------------------------------------------------------
  test = merge(test , mean.due.pca, all.x = TRUE, by = c("Producto_ID", "Cliente_ID", "Agencia_ID"))
  #-----------------------------------------------------------------
  test = merge(test , mean.vh.age, all.x = TRUE, by = c("Agencia_ID"))
  
  
  test = merge(test , sd.due.acrcp, all.x = TRUE, 
               by = c("Producto_ID", "Cliente_ID", 
                      "Agencia_ID",	"Canal_ID" , "Ruta_SAK"))
  
  test = merge(test , mean.due.acrcp, all.x = TRUE, 
               by = c("Producto_ID", "Cliente_ID", 
                      "Agencia_ID",	"Canal_ID" , "Ruta_SAK"))
  
  set.seed(100)
  temp = work.train.data.fun( test )
  train=temp[[1]]
  test3=temp[[2]]
  set.seed(100)
  #train,test3,nro,eta=0.1,md=10,cb=0.5
  value = xgb.fun(train,test3,75,0.1,8,0.5)
  (v1=value[[1]])
  
  model1=value[[2]]
  #----------------------------------------------------------------
  return(list(c(v1),model1))
}

bind.var.main.test.fun = 
		function(main.test,
				mean.due.age,
				mean.due.can,
				mean.due.rut ,
				mean.due.cli ,
				mean.due.pro ,
				mean.due.pa,
				mean.due.pr,
				mean.due.pcli,
				mean.due.pcan,
				mean.due.pca,
				mean.vh.age,
				mean.due.acrcp,
				sd.due.acrcp){

	main.test= merge(main.test, mean.due.age, all.x = TRUE, by = c("Agencia_ID"))
	main.test= merge(main.test, mean.due.can, all.x = TRUE, by = c("Canal_ID"))
	main.test= merge(main.test, mean.due.rut, all.x = TRUE, by = c("Ruta_SAK"))
	main.test= merge(main.test, mean.due.cli, all.x = TRUE, by = c("Cliente_ID"))
	#main.test= merge(main.test, mean.due.pro, all.x = TRUE, by = c("Producto_ID"))

	main.test= merge(main.test, mean.due.pa, all.x = TRUE, by = c("Producto_ID", "Agencia_ID"))
	main.test= merge(main.test, mean.due.pr, all.x = TRUE, by = c("Producto_ID", "Ruta_SAK"))
	main.test= merge(main.test, mean.due.pcli, all.x = TRUE, by = c("Producto_ID", "Cliente_ID"))
	main.test= merge(main.test, mean.due.pcan, all.x = TRUE, by = c("Producto_ID", "Canal_ID"))

	main.test= merge(main.test, mean.due.pca, all.x = TRUE, by = c("Producto_ID", "Cliente_ID", "Agencia_ID"))

	main.test= merge(main.test, mean.vh.age, all.x = TRUE, by = c("Agencia_ID"))

	main.test = merge(main.test , sd.due.acrcp, all.x = TRUE, 
		by = c("Producto_ID", "Cliente_ID", 
				"Agencia_ID",	"Canal_ID" , "Ruta_SAK"))

	main.test = merge(main.test , mean.due.acrcp, all.x = TRUE, 
		by = c("Producto_ID", "Cliente_ID", 
				"Agencia_ID",	"Canal_ID" , "Ruta_SAK"))

	return(main.test)
}

work.rfm.data.fun=function(main.train.x){
	#data 排序
	rfm.data0 =
	main.train.x  %>% 
	arrange(	Agencia_ID , Canal_ID , 
			Ruta_SAK , Cliente_ID , 
			Producto_ID , Semana)

	#算 rfm
	temp = summarise(
		group_by(
			rfm.data0, Agencia_ID , Canal_ID , 
			Ruta_SAK , Cliente_ID , Producto_ID) 
	,recent=8-max(Semana),freq=length(Semana),money=mean(Venta_hoy))

	rfm.data = data.table( temp )

	return(rfm.data)
}
freq.level.fun=function(x){
	#x=0
	if(x>=1 && x<=2){
		value=1
	}else if(x>=3 && x<=4){
		value=2			
	}else if(x>=5 && x<=6){
			value=3			
	}
	return(value)
}
recent.level.fun=function(x){
	#x=0
	if(x>=0 && x<=2){
		value=3
	}else if(x>=2 && x<=3){
		value=2			
	}else if(x>=4 && x<=5){
			value=1			
	}
	return(value)
}
money.level.fun=function(x){
	#x=0
	if(x>=0 && x<=19){
		value=1
	}else if(x>19 && x<=36){
		value=2			
	}else if(x>36 ){
			value=3			
	}
	return(value)
}


pred.fun=function(main.train.x,main.train.x.2,train.amount,main.test){
  #--------------------------------------------------------------------------------------------
  #生變數
  main.train.y$log.due	= log1p( main.train.y$Demanda_uni_equil )
  main.train.x = data.table( main.train.x )
  
  temp = work.var.fun(main.train.x)
  mean.due.age	=	temp[[1]]
  mean.due.can	=	temp[[2]]
  mean.due.rut	=	temp[[3]]
  mean.due.cli	=	temp[[4]]
  mean.due.pro	=	temp[[5]]
  mean.due.pa	=	temp[[6]]
  mean.due.pr	=	temp[[7]]
  mean.due.pcli	=	temp[[8]]
  mean.due.pcan	=	temp[[9]]
  mean.due.pca	=	temp[[10]]
  mean.vh.age	=	temp[[11]]
  mean.due.acrcp=	temp[[12]]
  sd.due.acrcp	=	temp[[13]]
  gc()
  #--------------------------------------------------------------------------------------------
  #
  sample.amount = nrow(main.train.y)
  main.train.y = data.table(main.train.y)
  temp2 = work.model.fun2(main.train.y,
                          train.amount,sample.amount,
                          mean.due.age,
                          mean.due.can,
                          mean.due.rut ,
                          mean.due.cli ,
                          mean.due.pro ,
                          mean.due.pa,
                          mean.due.pr,
                          mean.due.pcli,
                          mean.due.pcan,
                          mean.due.pca,
                          mean.vh.age,
                          mean.due.acrcp,
                          sd.due.acrcp)
  gc()
  print( temp2[[1]] )
  model=temp2[[2]]
  
  rm(temp2)
  rm(temp)
  #rm(main.train.x) 	
  gc()
  
  #-----------------------------------------------------------------
  #-----------------------------------------------------------------	
  #-----------------------------------------------------------------
  #模型生好了   接下來是預測
  
  #--------------------------------------------------------------------------------------------
  #生變數
  #main.train.y$log.due	= log1p( main.train.y$Demanda_uni_equil )
  main.train.x.2 = data.table(main.train.x.2)
  temp = work.var.fun(main.train.x.2)
  mean.due.age	=	temp[[1]]
  mean.due.can	=	temp[[2]]
  mean.due.rut	=	temp[[3]]
  mean.due.cli	=	temp[[4]]
  mean.due.pro	=	temp[[5]]
  mean.due.pa	=	temp[[6]]
  mean.due.pr	=	temp[[7]]
  mean.due.pcli	=	temp[[8]]
  mean.due.pcan	=	temp[[9]]
  mean.due.pca	=	temp[[10]]
  mean.vh.age	=	temp[[11]]
  mean.due.acrcp=	temp[[12]]
  sd.due.acrcp	=	temp[[13]]
  rm(temp)
  gc()
  #---------------------------------------------------
  #---------------------------------------------------
  main.test = bind.var.main.test.fun(main.test,
                                     mean.due.age,
                                     mean.due.can,
                                     mean.due.rut ,
                                     mean.due.cli ,
                                     mean.due.pro ,
                                     mean.due.pa,
                                     mean.due.pr,
                                     mean.due.pcli,
                                     mean.due.pcan,
                                     mean.due.pca,
                                     mean.vh.age,
                                     mean.due.acrcp,
                                     sd.due.acrcp)
  
  pred1<-predict(model,xgb.DMatrix(data.matrix(
    main.test[,c(8:ncol(main.test)),with=FALSE]),
    missing=NA))
  pred1.log = expm1( pred1 )
  pred1.log=as.integer(round(pred1.log))
  
  result1 = data.table(	id=main.test$id,
                        Demanda_uni_equil=pred1.log)
  
  return((result1))
}

File: function.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Mar 17 22:24:45 2018

@author: linsam
"""


import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import xgboost as xgb
from random import sample, seed
import copy
from datetime import datetime
import gc
#-----------------------------------------

def cbind(base,data):
    #base = train_y
    #data = log_due
    data.index = base.index
    bind_data = pd.concat([base, data], axis=1)
    
    return bind_data

def add_group_col(data):
    #data = mean_due_age
    name = data.index.name
    data[name]=data.index
    data.index  = range(len(data))
    
    return data



def change_type2cate(train_x):
    
    colname = ['Agencia_ID' , 'Canal_ID' , 'Ruta_SAK' , 
           'Cliente_ID' , 'Producto_ID'  ]
    for col in colname:
        train_x[col] = train_x[col].astype('category')
    
    return train_x
    
    #-----------------------------------------------

def feature_engineering(train_x):
    #s = datetime.datetime.now()

    #mean_due_age = 
    #mean.due.age = main.train.x[, .(mean.due.age = mean(log.due)), by = .(Agencia_ID)]
    s = datetime.now()
    mean_due_age = train_x.groupby(['Agencia_ID'], as_index=False)['log_due'].agg({'mean_due_age':np.mean})
    mean_due_can = train_x.groupby(['Canal_ID'], as_index=False)['log_due'].agg({'mean_due_can':np.mean})
    mean_due_rut = train_x.groupby(['Ruta_SAK'], as_index=False)['log_due'].agg({'mean_due_rut':np.mean})
    mean_due_cli = train_x.groupby(['Cliente_ID'], as_index=False)['log_due'].agg({'mean_due_cli':np.mean})
    mean_due_pro = train_x.groupby(['Producto_ID'], as_index=False)['log_due'].agg({'mean_due_pro':np.mean})

	#--------------------------------------------------------------------------------------------
    mean_vh_age = train_x.groupby(['Agencia_ID'], as_index=False)['log_vh'].agg({'mean_vh_age':np.mean})
	#--------------------------------------------------------------------------------------------
    mean_due_pa = train_x.groupby(['Producto_ID','Agencia_ID'], as_index=False)['log_due'].agg({'mean_due_pa':np.mean})
    mean_due_pr = train_x.groupby(['Producto_ID','Ruta_SAK'], as_index=False)['log_due'].agg({'mean_due_pr':np.mean})
    mean_due_pcli = train_x.groupby(['Producto_ID','Cliente_ID'], as_index=False)['log_due'].agg({'mean_due_pcli':np.mean})
    mean_due_pcan = train_x.groupby(['Producto_ID','Canal_ID'], as_index=False)['log_due'].agg({'mean_due_pcan':np.mean})
    
    mean_due_pca = train_x.groupby(['Producto_ID','Cliente_ID','Agencia_ID'], as_index=False)
    mean_due_pca = mean_due_pca['log_due'].agg({'mean_due_pca':np.mean})
	#--------------------------------------------------------------------------------------------
    s = datetime.now()
    mean_due_acrcp = train_x.groupby(['Agencia_ID','Canal_ID','Ruta_SAK','Cliente_ID','Producto_ID'], as_index=False)
    mean_due_acrcp = mean_due_acrcp['log_due'].agg({'mean_due_acrcp':np.mean})
    sd_due_acrcp = train_x.groupby(['Agencia_ID','Canal_ID','Ruta_SAK','Cliente_ID','Producto_ID'], as_index=False)
    sd_due_acrcp = sd_due_acrcp['log_due'].agg({'sd_due_acrcp':np.std})
    t = datetime.now() - s
    print(t) 
    # 0:00:04.097429
    
    tem = [ mean_due_age,
            mean_due_can,
            mean_due_rut ,
            mean_due_cli ,
            mean_due_pro ,
            mean_due_pa,
            mean_due_pr,
            mean_due_pcli,
            mean_due_pcan,
            mean_due_pca,	
            mean_vh_age,
            mean_due_acrcp,
            sd_due_acrcp]

    return tem
    
def data_preprocess(real_train_x):
    log_vh = pd.DataFrame( np.log1p( real_train_x['Venta_hoy'] ) )
    log_vh.columns = ['log_vh']
    
    real_train_x['log_vh'] = log_vh
    
    log_due = pd.DataFrame( np.log1p( real_train_x['Demanda_uni_equil'] ) )
    log_due.columns = ['log_due']
    
    real_train_x['log_due'] = log_due
    
    return real_train_x

def merge_feature(train_y,tem,date_name ):
    
    mean_due_age    = tem[0]
    mean_due_can    = tem[1]
    mean_due_rut    = tem[2]
    mean_due_cli    = tem[3]
    mean_due_pro    = tem[4]
    mean_due_pa     = tem[5]
    mean_due_pr     = tem[6]
    mean_due_pcli   = tem[7]
    mean_due_pcan   = tem[8]
    mean_due_pca    = tem[9]	
    mean_vh_age     = tem[10]
    mean_due_acrcp  = tem[11]
    sd_due_acrcp    = tem[12]
    
    if date_name == 'train':
        seed(100)
        # first, we build model by frac 0.5 data
        # that cost more less time
        #test = train_y.sample(frac=0.5,replace=False)    
        # in final build model, we use all data
        test = train_y
        colname = ['Agencia_ID' , 'Canal_ID' , 'Ruta_SAK' , 
                   'Cliente_ID' , 'Producto_ID'  , 
                   'Demanda_uni_equil', 'log_due']
        test = test[colname]
        
    elif date_name == 'test':
        test = train_y
        colname = ['Agencia_ID' , 'Canal_ID' , 'Ruta_SAK' , 
                   'Cliente_ID' , 'Producto_ID'  ]
    
    #len(test)#6895
    #-------------------------------------------------------------------------
    test = pd.merge(test, mean_due_age,how = 'left', on='Agencia_ID')
    
    test = pd.merge(test, mean_due_can,how = 'left', on='Canal_ID')
    test = pd.merge(test, mean_due_rut,how = 'left', on='Ruta_SAK')
    test = pd.merge(test, mean_due_cli,how = 'left', on='Cliente_ID')
    #-------------------------------------------------------------------------
    test = pd.merge(test , mean_due_pa, how = 'left', on = ["Producto_ID", "Agencia_ID"])
    test = pd.merge(test , mean_due_pr, how = 'left', on = ["Producto_ID", "Ruta_SAK"])
    test = pd.merge(test , mean_due_pcli, how = 'left', on = ["Producto_ID", "Cliente_ID"])
    test = pd.merge(test , mean_due_pcan, how = 'left', on = ["Producto_ID", "Canal_ID"])
    #-------------------------------------------------------------------------
    test = pd.merge(test , mean_due_pca, how = 'left', 
                    on = ["Producto_ID", "Cliente_ID", "Agencia_ID"])
    #-----------------------------------------------------------------
    test = pd.merge(test , mean_vh_age, how = 'left', on = ["Agencia_ID"])
    test = pd.merge(test , sd_due_acrcp, how = 'left', 
                    on = ["Producto_ID", "Cliente_ID", "Agencia_ID",	"Canal_ID" , "Ruta_SAK"])
                    
    test = pd.merge(test , mean_due_acrcp, how = 'left', 
                    on = ["Producto_ID", "Cliente_ID", "Agencia_ID",	"Canal_ID" , "Ruta_SAK"])
      
    colname = ['Agencia_ID' , 'Canal_ID' , 'Ruta_SAK' , 'Cliente_ID' , 'Producto_ID']
        
    test.drop(colname, axis=1, inplace=True)# inplace is update test (drop)  
    
    return test


def rmsle_eval(y, y0):
    
    y0=y0.get_label()    
    assert len(y) == len(y0)
    return 'rmsle',np.sqrt(np.mean(np.power(np.log1p(y)-np.log1p(y0), 2)))
    
    
def build_model(test):
    
    test_y = copy.deepcopy( test['log_due'] )
    test_x = copy.deepcopy( test )
    test_x.drop('log_due', axis=1, inplace=True)
    
    X_train, X_test, y_train, y_test = train_test_split(
        test_x, test_y, test_size=0.5, random_state=0)

    real_train_y = copy.deepcopy( X_train['Demanda_uni_equil'] )
    real_test_y = copy.deepcopy( X_test['Demanda_uni_equil'] )
    
    X_train.drop(['Demanda_uni_equil'], axis=1, inplace=True)
    X_test.drop(['Demanda_uni_equil'], axis=1, inplace=True)   

    dtrain = xgb.DMatrix(X_train,y_train)
    dtest = xgb.DMatrix(X_test,y_test)
    
    xgb_params={
            'eta':0.1,
            'max_depth':7,
            'subsample':0.7,
            'colsample_bytree':0.7,
            'objective':'reg:linear',
            'nfold':2,
            'eval_metric': 'rmse',
            'silent': 1
            ,'tree_method':'gpu_hist'
            }
    
    '''cv_output = xgb.cv(xgb_params,
                       dtrain,
                       num_boost_round=75,
                       early_stopping_rounds=5,
                       verbose_eval=5,
                       show_stdv=False)'''
    
    watchlist  = [ (dtrain,'train'),(dtest,'eval')]
                 
    try:
        del model
        gc.collect()
    except:
        123
      
    seed(100)
    #s = datetime.now()
    model = xgb.train(xgb_params, 
                      dtrain, 
                      evals = watchlist,
                      num_boost_round = 100,
                      early_stopping_rounds = 1,
                      verbose_eval = 1
                      #feval = rmsle_eval
                      )
    #t = datetime.now() -s 
    #print(t)


    #model.save_model('0001.model')
    #model = xgb.Booster({'nthread': 4})  # init model
    #model.load_model('model.bin')  # load data              
                      
    #------------------------------------------
    train_pred = model.predict(dtrain)
    train_pred = np.expm1(train_pred).astype(int)
    
    test_pred = model.predict(dtest)
    test_pred = np.expm1(test_pred).astype(int)
    
    temp = np.array( real_train_y )
    temp = np.log1p( temp ) -  np.log1p( train_pred )
    train_rmsle = np.sqrt( np.mean( np.power(temp, 2) ) )

    temp = np.array( real_test_y )
    temp = np.log1p( temp ) -  np.log1p( test_pred )
    test_rmsle = np.sqrt( np.mean( np.power(temp, 2) ) )

    print('train_rmsle :' + str( train_rmsle ) + '\ntest_rmsle :' + str( test_rmsle ) )
    
    ''' compare gpu vs cpu
    gpu cost time : 0:00:44.152872
    cpu cost time : 0:10:30.306069
    -------------------gpu------------------------                   
    train_rmsle :0.48128475735646753
    test_rmsle :0.4830263706408986
    kaggle Private Score    : 0.47381
    kaggle Public Score     : 0.45840
    -------------------cpu------------------------
    train_rmsle :0.4852542921969597
    test_rmsle :0.4867282065421481
    kaggle Private Score    : 0.47636
    kaggle Public Score     : 0.46137
    ===================conclusion==============
    gpu is faster 10 times than cpu, and gpu error is smaller than cpu
    end'''
    

    return model


def build_model2(test):
    
    test_y = copy.deepcopy( test['log_due'] )
    test_x = copy.deepcopy( test )
    test_x.drop(['log_due'], axis=1, inplace=True)
    
    X_train, X_test, y_train, y_test = train_test_split(
        test_x, test_y, test_size=0.5, random_state=0)
    
    real_train_y = copy.deepcopy( X_train['Demanda_uni_equil'] )
    real_test_y = copy.deepcopy( X_test['Demanda_uni_equil'] )
    
    X_train.drop(['Demanda_uni_equil'], axis=1, inplace=True)
    X_test.drop(['Demanda_uni_equil'], axis=1, inplace=True)    
    
    #dtrain = xgb.DMatrix(X_train,y_train)
    #dtest = xgb.DMatrix(X_test,y_test)    
    
    eval_set = [(X_train, y_train), (X_test, y_test)]
    
    xgb_param = {'max_depth':7, 
                 'subsample':0.7,
                 'colsample_bytree':0.7, 
                 'learning_rate':0.1
                 ,'tree_method':'gpu_hist'
                 ,'max_bin' : 16
                 }    
    try:
        del model
        gc.collect()
    except:
        123
        
    model = xgb.XGBRegressor(**xgb_param, n_estimators=100)  
     
    seed(100)
    model.fit(X_train,y_train,
              eval_set = eval_set,
              verbose=True,
              eval_metric = ['rmse']
              #,early_stopping_rounds = 1
              )
    #model.save_model('0001.model')
    #model = xgb.Booster({'nthread': 4})  # init model
    #model.load_model('model.bin')  # load data
    #------------------------------------------
    train_pred = model.predict(X_train)
    train_pred = np.expm1(train_pred).round().astype(int)
    
    test_pred = model.predict(X_test)
    test_pred = np.expm1(test_pred).round().astype(int)
    
    temp = np.array( real_train_y )
    temp = np.log1p( temp ) -  np.log1p( train_pred )
    train_rmsle = np.sqrt( np.mean( np.power(temp, 2) ) )

    temp = np.array( real_test_y )
    temp = np.log1p( temp ) -  np.log1p( test_pred )
    test_rmsle = np.sqrt( np.mean( np.power(temp, 2) ) )

    print('train_rmsle :' + str( train_rmsle ) + '\ntest_rmsle :' + str( test_rmsle ) )
    
    return model,train_rmsle,test_rmsle


def plot_fun(model):
    
    results = model.evals_result()
    
    from matplotlib import pyplot
    epochs = len(results['validation_0']['rmse'])
    x_axis = range(0, epochs)
    
    fig, ax = pyplot.subplots()
    ax.plot(x_axis, results['validation_0']['rmse'], label='Train')
    ax.plot(x_axis, results['validation_1']['rmse'], label='Test')

    ax.legend()
    pyplot.ylabel('rmse')
    pyplot.title('XGBoost rmse')
    pyplot.show()    









